cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -684,6 +685,85 @@ Result<S3Model::GetObjectResult> GetObjectRange(Aws::S3::S3Client* client,
   return OutcomeToResult(client->GetObject(req));
 }
 
+template <typename ObjectResult>
+std::shared_ptr<const KeyValueMetadata> GetObjectMetadata(const ObjectResult& result) {
+  std::vector<std::string> keys, values;
+
+  auto push = [&](std::string k, const Aws::String& v) {
+    if (!v.empty()) {
+      keys.push_back(std::move(k));
+      values.push_back(FromAwsString(v).to_string());
+    }
+  };
```

Comment:
You can create a mutable KeyValueMetadata and return it from this function (`shared_ptr<T>` implicitly converts to `shared_ptr<const T>`)
```suggestion
  std::shared_ptr<KeyValueMetadata> metadata;

  auto push = [&](std::string k, const Aws::String& v) {
    if (!v.empty()) {
      metadata->Append(std::move(k), FromAwsString(v).to_string());
    }
  };
```

Reply:
Hmm, why not.